### PR TITLE
#112: Add server-client protocol and feature matrix

### PR DESCRIPTION
**Pull request by martin-fleck-at**
Part of https://github.com/eclipse-glsp/glsp/issues/112

Signed-off-by: Martin Fleck <mfleck@eclipsesource.com>

### CODE REVIEW

File: PROTOCOL.md [hunk 1]
```diff
@@ -0,0 +1,2296 @@
...
+    public readonly markers: Marker[];
+}
+```
+</details>
+
+## 2.12. Element Navigation
+
+GLSP makes no assumption about the type of navigation a user may want to perform. Thus a generic infrastructure is provided that the client and server can use to implement specific navigation types, e.g., navigation to documentation, implementation, etc. The type of navigation is identified by the `targetTypeId`.
+
+A client may request the targets for a specific type of navigation by querying the server to which the server will respond with a set of navigation targets. A `NavigationTarget` identifies the object we want to navigate to via it's uri and may further provide a label to display for the client. Additionally, generic  arguments may be used to to encode any domain- or navigation type-specific information.
```
Comment (planger) [inline PROTOCOL.md:1427]:
```suggestion
A client may request the targets for a specific type of navigation by querying the server to which the server will respond with a set of navigation targets. A `NavigationTarget` identifies the object we want to navigate to via its uri and may further provide a label to display for the client. Additionally, generic arguments may be used to to encode any domain- or navigation type-specific information.
```

File: PROTOCOL.md [hunk 2]
```diff
@@ -0,0 +1,2296 @@
...
+     * Additional arguments for custom behavior.
+     */
+    public readonly args?: Args;
+}
+```
+</details>
+
+### 2.17.3. ApplyLabelEditOperation
+
+A very common use case in domain models is the support of labels that display textual information to the user. For instance, the `SGraph` model of Sprotty has support for labels that can be attached to a node, edge, or port, and that contain some text that is rendered in the view. To apply new text to such a label element the client may send a `ApplyLabelEditOperation` to the server.
```
Comment (planger) [inline PROTOCOL.md:1981]:
```suggestion
A very common use case in domain models is the support of labels that display textual information to the user. For instance, the `SGraph` model of Sprotty has support for labels that can be attached to a node, edge, or port, and that contain some text that is rendered in the view. To apply new text to such a label element the client may send an `ApplyLabelEditOperation` to the server.
```

File: PROTOCOL.md [hunk 3]
```diff
@@ -0,0 +1,2296 @@
...
+     * Text that should be applied on the label.
+     */
+    public readonly text: string;
+}
+```
+</details>
+
+## 2.18. Clipboard
+
+In GLSP the clipboard needs to be managed by the client but the conversion from data to a clipboard-compatible format is handled by the server. By default, we use `application/json` as exchange format.
```
Comment (planger) [inline PROTOCOL.md:2007]:
```suggestion
In GLSP the clipboard needs to be managed by the client but the conversion from the selection to be copied into a clipboard-compatible format is handled by the server. By default, GLSP use `application/json` as exchange format.
```

File: PROTOCOL.md [hunk 4]
```diff
@@ -0,0 +1,2296 @@
...
+```
+</details>
+
+### 2.20.3. Context Menu
+
+The context menu is an overlay that is triggered by a right click from the user. The menu may be filled with actions from the client but may also be filled with actions from the server. If server actions are to be used, the client needs to send a `RequestContextActions` action with context id `context-menu` and handle the returned actions from the `SetContextActions` response accordingly, e.g., rendering them in a context menu.
+
+### 2.20.4. Command Palette
+
+The command palette is an "auto-complete" widget that is triggered the user hitting `ctrl+space`. The menu may be filled with actions from the client but may also be filled with actions from the server. If server actions are to be used, the client needs to send a `RequestContextActions` action with context id `command-palette` and handle the returned actions from the `SetContextActions` response accordingly, e.g., rendering them in a auto-complete widget.
```
Comment (planger) [inline PROTOCOL.md:2207]:
```suggestion
The command palette is an "auto-complete" widget that is triggered when the user hits `Ctrl+Space`. The menu may be filled with actions from the client but may also be filled with actions from the server. If server actions are to be used, the client needs to send a `RequestContextActions` action with context id `command-palette` and handle the returned actions from the `SetContextActions` response accordingly, i.e., rendering them in a auto-complete widget.
```

**Review by planger — CHANGES_REQUESTED** [review]
Excellent, thanks a lot! I made a few suggestions inline... If you are fine with them, please confirm and we can merge this change. Thanks!

**Comment by martin-fleck-at** [comment]
@planger Thank you for your fast review and the very valuable suggestions! I agree with every change you proposed and applied all suggestions in a follow-up commit to this PR. If you have no further comments and you are happy with the feature matrix as well, you can merge the change. Thank you!

**Review by planger — APPROVED** [review]

**Pull request merged**